src/OrchardCore/OrchardCore.ContentManagement.Abstractions/ContentExtensions.cs

```diff
@@ -65,7 +65,10 @@ public static ContentElement Get(this ContentElement contentElement, Type conten
         {
             if (contentElement.Elements.TryGetValue(name, out var element))
             {
-                return element;
+                if (element.GetType() == contentElementType)
```

Comment:
I don't think we should guard here against invalid usage. I would say make sure you are calling the method correctly to begin with

Reply:
I think the idea is not that bad, but we could instead allow deserializing a more specific type here, e.g. something like this:

```C#
if (contentElement.Elements.TryGetValue(name, out var element))
{
    var currentType = element.GetType();

    if (currentType == contentElementType ||
        !currentType.IsAssignableFrom(contentElementType))
        return element;
}
[...]
```

Reply:
maybe without the or?

Reply:
Hmm, shouldn't it immediately throw if the type is invalid?

Reply:
The main problem here is that the invalid use will not always cause the serialize and not take advantage of the cache. Someone **unknowingly** can be following invalid use and not take advantage of caching.

Reply:
Yes, my idea was just to allow derived types to deserialize again, otherwise either return the previously deserialized instance (as before) or (maybe better) throw an `InvalidCastException` like the generic `Get<T>()` method.

Reply:
I don't have strong opinion here. Personally, I rather not encourage bad practice. @Piedone @hishamco thoughts on this PR?

Reply:
Combine the condition with the `TryGetValue`, and use `IsAssignableTo` instead of type==type such that base types would work, e.g. calling BlogPart THEN ContentPart would then work as an instance of BlogPart "is assignable to" ContentPArt.

Add the unit test from @gvkries and another one that does the BlogPart "then" ContentPart calls.

Reply:
I agree with that.

An alternative can be to use type+name as the `Elements` key. However, this would require changes to all the other `Elements` accesses in the class. And I'm not sure if the frequency of the use case (i.e. `Get`ting the same content part of the same content item object with different types) warrants anything more complex.